diff --git a/src/core/web_contents_delegate_qt.cpp b/src/core/web_contents_delegate_qt.cpp
index 2dd75df..7476e47 100644
--- a/src/core/web_contents_delegate_qt.cpp
+++ b/src/core/web_contents_delegate_qt.cpp
@@ -258,7 +258,14 @@ void WebContentsDelegateQt::DidFinishLoad(content::RenderFrameHost* render_frame
         m_viewClient->iconChanged(QUrl());
 
     m_viewClient->loadProgressChanged(100);
-    m_viewClient->loadFinished(true, toQt(validated_url));
+
+    content::NavigationEntry *entry = web_contents()->GetController().GetActiveEntry();
+    if (!entry) {
+        m_viewClient->loadFinished(true, toQt(validated_url));
+        return;
+    }
+    int http_statuscode = entry->GetHttpStatusCode();
+    m_viewClient->loadFinished(true, toQt(validated_url), false, http_statuscode);
 }
 
 void WebContentsDelegateQt::DidUpdateFaviconURL(const std::vector<content::FaviconURL> &candidates)
diff --git a/src/webengine/api/qquickwebengineview.cpp b/src/webengine/api/qquickwebengineview.cpp
index 1b1dcec..016093c 100644
--- a/src/webengine/api/qquickwebengineview.cpp
+++ b/src/webengine/api/qquickwebengineview.cpp
@@ -491,7 +491,7 @@ void QQuickWebEngineViewPrivate::loadFinished(bool success, const QUrl &url, boo
     }
     if (success) {
         explicitUrl = QUrl();
-        QQuickWebEngineLoadRequest loadRequest(url, QQuickWebEngineView::LoadSucceededStatus);
+        QQuickWebEngineLoadRequest loadRequest(url, QQuickWebEngineView::LoadSucceededStatus, errorDescription, errorCode);
         Q_EMIT q->loadingChanged(&loadRequest);
         return;
     }
