diff --git a/src/core/web_engine_context.cpp b/src/core/web_engine_context.cpp
index 39e11a9..244ae09 100644
--- a/src/core/web_engine_context.cpp
+++ b/src/core/web_engine_context.cpp
@@ -358,10 +358,12 @@ WebEngineContext::WebEngineContext()
         }
     }
 
+    // ANGLE is said to cause threading and performance issues when not disabling all GPU features,
+    // though force OpenGL ES mode Chromium-side and let's see what happens..
     if (glType)
         parsedCommandLine->AppendSwitchASCII(switches::kUseGL, glType);
     else
-        parsedCommandLine->AppendSwitch(switches::kDisableGpu);
+        parsedCommandLine->AppendSwitchASCII(switches::kUseGL, gfx::kGLImplementationEGLName);
 
     content::UtilityProcessHostImpl::RegisterUtilityMainThreadFactory(content::CreateInProcessUtilityThread);
     content::RenderProcessHostImpl::RegisterRendererMainThreadFactory(content::CreateInProcessRendererThread);
