diff --git a/src/codec/AVDecoder.cpp b/src/codec/AVDecoder.cpp
index 4c3fce6..687cea2 100644
--- a/src/codec/AVDecoder.cpp
+++ b/src/codec/AVDecoder.cpp
@@ -76,11 +76,16 @@ bool AVDecoder::open()
     const QString hwa = property("hwaccel").toString();
     AVCodec* codec = get_codec(codecName(), hwa, d.codec_ctx->codec_id);
     if (!codec) { // TODO: can be null for none-ffmpeg based decoders
-        QString es(tr("No codec could be found for '%1'"));
+        QString es = QStringLiteral("No codec could be found for '%1'");
         if (d.codec_name.isEmpty()) {
-            es = es.arg(QLatin1String(avcodec_get_name(d.codec_ctx->codec_id)));
-            if (!hwa.isEmpty())
-                es.append('_').append(hwa);
+            const char* tmpCodecName = avcodec_get_name(d.codec_ctx->codec_id);
+            if (!tmpCodecName) {
+                es = es.arg(QStringLiteral("unknown"));
+            } else {
+                es = es.arg(QString(tmpCodecName));
+                if (!hwa.isEmpty())
+                    es.append('_').append(hwa);
+            }
         } else {
             es = es.arg(d.codec_name);
         }
