diff --git a/src/codec/video/VideoDecoderDXVA.cpp b/src/codec/video/VideoDecoderDXVA.cpp
index 64b12f3..a325982 100644
--- a/src/codec/video/VideoDecoderDXVA.cpp
+++ b/src/codec/video/VideoDecoderDXVA.cpp
@@ -314,9 +314,16 @@ QVector<GUID> VideoDecoderDXVAPrivate::getSupportedCodecs() const
     UINT input_count = 0;
     GUID *input_list = NULL;
     QVector<GUID> guids;
+
     DX_ENSURE_OK(vs->GetDecoderDeviceGuids(&input_count, &input_list), guids);
-    guids.resize(input_count);
-    memcpy(guids.data(), input_list, input_count*sizeof(GUID));
+
+    if (input_count < 1 || !input_list)
+        return guids;
+
+    for (UINT i = 0; i < input_count; ++i) {
+        guids.append(input_list[i]);
+    }
+
     CoTaskMemFree(input_list);
     return guids;
 }
